src/Xrefcheck/RepoInfo.hs

```diff
@@ -0,0 +1,119 @@
+{- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+{-# LANGUAGE GADTs #-}
+
+module Xrefcheck.RepoInfo
```

Comment:
I have solved the file system casing problem with:

- The weaker custom canonizalisation.
- Implementing `RepoInfo` in a way that it allows to search in a case-insensitive way.

This could have been implemented by carrying boolean values together with the `Map` keys telling if they are case-insensitive or not, but then we should have been taking into account cases that should not be possible, such as comparing a case-sensitive key with a case-insensitive one.

I have implemented it with a GADT instead, trying to hide the implementation and exposing a clear interface.

Reply:
First of all, have you checked how GitHub and GitLab handle the casing of filepaths? Does it for sure work similarly to anchors handling?

---

On the GADT / existential quantification, looks neat.

I suppose there were multiple ways to adopt our code here. The clue on which option to choose can give the following question: If we add more scanners (e.g. for Haskell sources), and there references have a different format, will it work?

Given that, we probably want to carry not a sum type with `RepoInfoData`s - either case-sensitive, or case-insensitive, rather multiple `RepoInfoData`, because in the future we can have several scanners that assume different filepath format and different comparison rules.

So I'd propose the following changes:

1. As you written, we will have `newtype`s over `FilePath` to indicate its format, casing, e.t.c.
     * Each such newtype will probably have to implement `Ord` instance + a typeclass providing a way to convert `FilePath` (platform-dependent path pointing to a previously scanned file in the filesystem) to this newtype. So we will probably do conversion not from Markdown to System filepath, but vice-versa (that's a debatable point!)
     
2. Let `Reference`, `FileInfo` probably some other types that use them be polymorphic over the format of filepath. We will let different scanners return different formats of links.
    * This will require `Reference` to keep a sum type for link (local file or external link), but this seems not that bad.
    * Our Markdown scanner will return case-sensitive or case-insensitive markdown links.

3. `RepoInfo` will keep a map from type to `RepoInfoData <this type>`.

    I.e. I should be able to write `lookupRepoInfoData (Proxy @CaseInsensitivePath)` and get `RepoInfoData CaseInsensitivePath`. We can use [dependent-map](https://hackage.haskell.org/package/dependent-map-0.4.0.0/docs/Data-Dependent-Map.html) or plain `Map TypeRep SomeRepoInfoData` here.

    Depending on the reference that we are currently checking, we will extract the map with the appropriate newtype.

Good that you extracted `RepoInfo` to a separate module, this will help us to introduce an appropriate interface in a better way.

Reply:
🤦 I tried how anchors behave in both of them, but not filepaths. It turns out that GitHub, which is case-insensitive regarding anchors, is case-sensitive regarding filepaths.

I think that it is better to undo all this `RepoInfo` part for avoiding unnecessary complexity, right? It gets solved with just the weaker custom canonizalisation version.

Reply:
Yes, let's undo the `RepoInfo` extraction.

But please preserve it some way, probably: move the essence of those commits to a separate branch and create a Draft pull request, then drop a link to it in this thread. I believe this extraction will likely be useful in the future.

Reply:
Done!

https://github.com/serokell/xrefcheck/pull/243